test/integration/model/ResultsServiceTest.php

```diff
@@ -270,71 +255,73 @@ public function testGetTestTaker()
-        $impProphecy = $prophet->prophesize('oat\taoOutcomeRds\model\RdsResultStorage');
-
+        $impProphecy = $this->prophesize(RdsResultStorage::class);
         $impProphecy->deleteResult('#foo')->willReturn(true);
         $imp = $impProphecy->reveal();
+
         $this->service->setImplementation($imp);
 
         $this->assertTrue($this->service->deleteResult('#foo'));
     }
 
     /**
+     * ResultService refactoring is required.
      *
      * @author Lionel Lecaque, lionel@taotesting.com
-     *         @expectedException \common_exception_Error (@todo fix)
-     *         @expectedExceptionMessage This delivery has no Result Server
      */
     public function testGetReadableImplementationNoResultStorage()
     {
-        $prophet = new Prophet();
+        // TODO: Refactor ResultService class to be able to mock dependencies and fix test.
+        $this->markTestSkipped();
+
+        $this->expectException(\common_exception_Error::class);
 
-        $deliveryProphecy = $prophet->prophesize('core_kernel_classes_Resource');
+        $deliveryProphecy = $this->prophesize(core_kernel_classes_Resource::class);
         $delivery = $deliveryProphecy->reveal();
 
         $this->service->getReadableImplementation($delivery);
     }
 
     /**
+     * ResultService refactoring is required.
      *
      * @author Lionel Lecaque, lionel@taotesting.com
-     *         @expectedException \common_exception_Error (@todo fix)
-     *         @expectedExceptionMessage This delivery has no readable Result Server
      */
     public function testGetReadableImplementationNoResultServer()
     {
-        $prophet = new Prophet();
+        // TODO: Refactor ResultService class to be able to mock dependencies and fix test.
+        $this->markTestSkipped();
 
-        $resultProphecy = $prophet->prophesize('core_kernel_classes_Resource');
+        $this->expectException(\common_exception_Error::class);
 
+        $resultProphecy = $this->prophesize(core_kernel_classes_Resource::class);
         $resultServer = $resultProphecy->reveal();
 
-        $deliveryProphecy = $prophet->prophesize('core_kernel_classes_Resource');
-
+        $deliveryProphecy = $this->prophesize(core_kernel_classes_Resource::class);
         $deliveryProphecy->getOnePropertyValue(new \core_kernel_classes_Property(DeliveryContainerService::PROPERTY_RESULT_SERVER))->willReturn($resultServer);
-
         $delivery = $deliveryProphecy->reveal();
+
         $this->service->getReadableImplementation($delivery);
     }
 
     /**
      *
      * @author Lionel Lecaque, lionel@taotesting.com
-     *         @expectedException \common_exception_Error (@todo fix)
-     *         @expectedExceptionMessage This delivery has no readable Result Server
      */
     public function testGetReadableImplementationNoResultServerModel()
     {
-        $prophet = new Prophet();
+        // TODO: Refactor ResultService class to be able to mock dependencies and fix test.
```

Comment:
I don't see why we aren't able to mock trait functions the same way as we mock class functions, could you please explain?

Reply:
One of problems here is that ResultService class uses oat\tao\model\ClassServiceTrait trait. There is no method to set ServiceLocator/Manager in this trait, therefore we can not mock any service in tests. It means in test we are always using real storage, file system etc.
Another difficulty here is because of implementation of some of internal methods, like ResultService::getItemIndexer(), ResultService::getDecompiledIndexer() for example. They use static methods and singletons, which make it's really difficult to test.
Of course we can mock this methods in ResultService class and run tests on this mocked object, but it won't be real test any more.

Reply:
You can mock getServiceManager with partial mock objects:

`
class A {
    public function aa(){
        var_dump($this->bb());
    }
    public function bb() {
        return 'FOOOO';
    }
}
`

`
       $aMock = $this->getMockBuilder(A::class)
            ->setMethods(['bb'])
            ->getMock();

        $aMock->method('bb')
            ->willReturn('BAR');

        $aMock->aa();
`

It prints `BAR`

Reply:
Concerning static methods and singletons - if that's a blocker (I'm not familiar), then ok, lets leave it...

Reply:
Yes you are right, I can mock service manager like this or use ServiceLocatorAwareTrait. 
I even tried it. But the biggest problem is usage of static methods, singletons and filesystem:
```
tao_models_classes_service_ServiceCallHelper::getInputValues();
tao_models_classes_service_FileStorage::singleton();
```

Reply:
Yes, you're right at this point, though we could forge static methods, I would prefer not to bring bad practices to the tests and wait for the code refactoring

It seems like all the discussions are done, so we are ready for this PR to be merged, right?

Reply:
I think yes